src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -294,6 +295,14 @@ public MemorySegment withOwnerThread(Thread newOwner) {
         }
     }
 
+    @Override
+    public void registerCleaner(Cleaner cleaner) {
+        checkAccessModes(CLOSE);
+        checkValidState();
+        MemoryScope.CleanupAction cleanupAction = scope.cleanupAction;
+        cleaner.register(this.scope, cleanupAction::cleanup);
```

Comment:
If `CleanupAction` extended from `Runnable` you could pass it in directly. e.g rename `cleanup` to `run`, and flip the two methods on `BasicCleanupAction`